### PR TITLE
UI - Project allocation

### PR DESCRIPTION
**Pull request by surekhag**

### CODE REVIEW

File: client/src/components/Employee/Employee.js [hunk 1]
```diff
@@ -63,15 +62,15 @@ const Employee = props => {
   // Load all emp info
   useEffect(() => {
     dispatch(loadManagers())
-  }, [])
+  }, [dispatch])
```
Comment (AatishBondge) [inline client/src/components/Employee/Employee.js:65]: I think dispatch is not going to change, you can remove this dependency.

File: client/src/components/Employee/Employee.js [hunk 2]
```diff
@@ -90,7 +89,7 @@ const Employee = props => {
       userForm.current.reset()
       dispatch(clearUserStatus())
     }
-  }, [addNewUserStatus, updateUserStatus, addToast])
+  }, [addNewUserStatus, updateUserStatus, addToast, dispatch])
```
Comment (AatishBondge) [inline client/src/components/Employee/Employee.js:92]: I think dispatch is not going to change, you can remove this dependency.

File: client/src/components/Employee/Employee.js [hunk 3]
```diff
@@ -100,15 +99,17 @@ const Employee = props => {
...
   const getStates = value => {
     if (value === null) return []
 
     const states = countryData.filter(item => {
-      if (item.country == value) {
+      if (item.country === value) {
         return item
       }
+
+      // return false;
```
Comment (AatishBondge) [inline client/src/components/Employee/Employee.js]: remove unwanted commented code

File: client/src/components/Project/Project.js [hunk 4]
```diff
@@ -262,14 +262,25 @@ const Project = props => {
...
                   ) : (
-                    <Button
-                      id="add"
-                      type="submit"
-                      color="primary"
-                      disabled={isSubmitting}
-                    >
-                      ADD PROJECT
-                    </Button>
+                    <>
```
Comment (AatishBondge) [inline client/src/components/Project/Project.js:239]: I think you should not write component file more than 200 lines, break it into sub component whenever possible.

File: client/src/api/employeeApi.js [hunk 5]
```diff
@@ -13,5 +13,5 @@ export function loadManagers() {
 }
 
 export function deleteEmployeeApi(id) {
-  return axios.delete(DELETE_USER_URL + id)
+  return axios.delete(USER_URL + id)
```
Comment (AatishBondge) [inline client/src/api/employeeApi.js]: use template literals instead Concatenation.

File: client/src/api/projectsApi.js [hunk 6]
```diff
@@ -1,25 +1,43 @@
...
+  PROJECT_URL
 } from '../configurations/endPoints'
 
 export function loadAllProjectsApi() {
   return axios.get(FETCH_ALL_PROJECTS)
 }
 
 export function deleteProjectApi(id) {
-  return axios.delete(DELETE_PROJECT_URL + id)
+  return axios.delete(PROJECT_URL + id)
```
Comment (AatishBondge) [inline client/src/api/projectsApi.js]: use template literals instead Concatenation.

File: client/src/api/projectsApi.js [hunk 7]
```diff
@@ -1,25 +1,43 @@
...
+  return axios.post(ALLOCATE_NEW_PROJECT_URL, reqBody)
+}
+
+export function projectAllocationDataApi(id) {
+  return axios.get(ALLOCATE_PROJECT_URL + id)
+}
+
+export function deallocateProjectApi(payload) {
+  const { enddate, id } = payload
+  const reqBody = { enddate: enddate }
```
Comment (AatishBondge) [inline client/src/api/projectsApi.js]: here if key and value are same then you can use direct key name only 
const reqBody = { enddate }

File: client/src/components/Project/ProjectAllocation.js [hunk 8]
```diff
@@ -169,205 +139,109 @@ const ProjectAllocation = props => {
...
+                        onChange={handleChange}
+                        disabledName="None"
+                        label={'Employee *'}
+                        value={values.employee}
+                      >
+                        {activeEmployees
+                          ? activeEmployees.map(item => {
+                              return (
+                                <MenuItem value={item._id}>
+                                  {item.firstname + ' ' + item.lastname}
```
Comment (AatishBondge) [inline client/src/components/Project/ProjectAllocation.js]: use template literals.

File: client/src/components/Project/ProjectAllocation.js [hunk 9]
```diff
@@ -169,205 +139,109 @@ const ProjectAllocation = props => {
...
+                    <GridItem xs={6} sm={6} md={6}>
+                      <SelectMenu
+                        name="project"
+                        onChange={handleChange}
+                        disabledName="Select Project"
+                        label={'Project *'}
+                        value={values.project}
+                      >
+                        {projects
+                          ? projects.map((item, key) => {
```
Comment (AatishBondge) [inline client/src/components/Project/ProjectAllocation.js]: if you want expression to render on true condition you can simply remove ? and add && . no need to return null
projects && {expression}

File: client/src/components/Project/ProjectAllocation.js [hunk 10]
```diff
@@ -169,205 +139,109 @@ const ProjectAllocation = props => {
...
-                      </div>
+                      <SelectMenu
+                        name="functional_manager"
+                        onChange={handleChange}
+                        disabledName="None"
+                        label={'Reporting Manager *'}
+                        value={values.functional_manager}
+                      >
+                        {managers
+                          ? managers.map(item => {
```
Comment (AatishBondge) [inline client/src/components/Project/ProjectAllocation.js]: if you want expression to render on true condition you can simply remove ? and add && . no need to return null
managers && {expression}

File: client/src/components/Project/ProjectAllocation.js [hunk 11]
```diff
@@ -169,205 +139,109 @@ const ProjectAllocation = props => {
...
+                              return (
+                                <MenuItem value={item._id}>
+                                  {item.firstname + ' ' + item.lastname}
```
Comment (AatishBondge) [inline client/src/components/Project/ProjectAllocation.js]: use template literals

File: client/src/components/Project/ProjectListing.js [hunk 12]
```diff
@@ -94,34 +94,39 @@ const ProjectListing = props => {
...
         enddate
       }
-      projectDetails.push(Object.values(data))
-      return
+      return Object.values(data)
     })
-    // Get/Delete data from Active projects
-    //  projectData = filteredProject;
+    projectDetails = [...tempProjectData]
+    // to do  - Get/Delete data from Active projects api
```
Comment (AatishBondge) [inline client/src/components/Project/ProjectListing.js:100]: remove comments

File: client/src/components/Project/updateProjectAllocation.js [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+  const deallocateProjectErr = useSelector(deallocateProjectError)
+  const delProjectAllocationStatus = useSelector(delProjectAllocationSuccess)
+  const delProjectAllocationErr = useSelector(delProjectAllocationError)
+
+  const [projectAllocations, setProjectAllocations] = useState([])
+  const [showDelDialog, setShowDelDialog] = useState(false)
+  const [projectAllocationToUpdate, setProjectAllocationToUpdate] = useState(
+    null
+  )
+  const project_id = projectToUpdate ? projectToUpdate[0]._id : null
```
Comment (AatishBondge) [inline client/src/components/Project/updateProjectAllocation.js]: use camel case variable names

File: client/src/components/Table/Table.js [hunk 14]
```diff
@@ -54,6 +57,10 @@ export default function CustomTable(props) {
               return (
                 <TableRow key={k} className={classes.tableBodyRow}>
                   {prop.map((prop, key) => {
```
Comment (AatishBondge) [inline client/src/components/Table/Table.js:59]: same name of array as well element. increases confusion.

File: client/src/components/Table/Table.js [hunk 15]
```diff
@@ -54,6 +57,10 @@ export default function CustomTable(props) {
...
+                    //This is only for deallocation of project where need _id in data but need not to show in UI.
+                    if (deallocateProject && key == 5) {
+                      return
```
Comment (AatishBondge) [inline client/src/components/Table/Table.js:62]: empty return ?

File: client/src/sagas/watchers/projectAllocation.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  deallocateProjectApi,
+  deleteProjectAllocationApi
+} from '../../api/projectsApi'
+
+function* workerAllocateProjectSaga(projectInfo) {
+  try {
+    const allocateProjectResponse = yield call(allocateProjectApi, projectInfo)
+
+    yield put(setAllocateProjectSuccess(allocateProjectResponse.data.message))
+    // yield put(loadAllProjects())
```
Comment (AatishBondge) [inline client/src/sagas/watchers/projectAllocation.js]: remove comments

File: server/app/api/models/project_allocation.js [hunk 17]
```diff
@@ -3,31 +3,33 @@
 const mongoose = require("mongoose");
 const Schema = mongoose.Schema;
```
Comment (AatishBondge) [inline server/app/api/models/project_allocation.js:4]: camelcase schema

File: client/src/components/Employee/Employee.js (hunk 3 above)
Comment (surekhag) [reply client/src/components/Employee/Employee.js]: Done

File: client/src/components/Employee/Employee.js (hunk 2 above)
Comment (surekhag) [reply client/src/components/Employee/Employee.js:92]: Wont fix.

File: client/src/components/Employee/Employee.js (hunk 1 above)
Comment (surekhag) [reply client/src/components/Employee/Employee.js:65]: Wont fix

File: client/src/components/Project/Project.js (hunk 4 above)
Comment (surekhag) [reply client/src/components/Project/Project.js:239]: Possible code moved to separate file to reduce current component total lines of code.

File: client/src/api/projectsApi.js (hunk 7 above)
Comment (surekhag) [reply client/src/api/projectsApi.js]: Done

File: client/src/components/Project/ProjectAllocation.js (hunk 8 above)
Comment (surekhag) [reply client/src/components/Project/ProjectAllocation.js]: Done

File: client/src/components/Project/ProjectAllocation.js (hunk 9 above)
Comment (surekhag) [reply client/src/components/Project/ProjectAllocation.js]: Done.

File: client/src/components/Project/ProjectAllocation.js (hunk 10 above)
Comment (surekhag) [reply client/src/components/Project/ProjectAllocation.js]: Done.

File: client/src/components/Project/ProjectAllocation.js (hunk 11 above)
Comment (surekhag) [reply client/src/components/Project/ProjectAllocation.js]: Done.

File: client/src/components/Project/ProjectListing.js (hunk 12 above)
Comment (surekhag) [reply client/src/components/Project/ProjectListing.js:100]: This is yet to be implemented ..

File: client/src/components/Project/updateProjectAllocation.js (hunk 13 above)
Comment (surekhag) [reply client/src/components/Project/updateProjectAllocation.js]: Done.

File: client/src/components/Table/Table.js (hunk 14 above)
Comment (surekhag) [reply client/src/components/Table/Table.js:59]: This is Material UI code. We can work on it separately.

File: client/src/components/Table/Table.js (hunk 15 above)
Comment (surekhag) [reply client/src/components/Table/Table.js:62]: yes. Dont want to code get executed based on condition mention. Same is also mention in comment line above.

File: client/src/sagas/watchers/projectAllocation.js (hunk 16 above)
Comment (surekhag) [reply client/src/sagas/watchers/projectAllocation.js]: Done.

File: server/app/api/models/project_allocation.js (hunk 17 above)
Comment (surekhag) [reply server/app/api/models/project_allocation.js:4]: This is syntax for Schema. Uppercase 'S'

File: client/src/api/employeeApi.js (hunk 5 above)
Comment (surekhag) [reply client/src/api/employeeApi.js]: Done for all apis.

File: client/src/api/projectsApi.js (hunk 6 above)
Comment (surekhag) [reply client/src/api/projectsApi.js]: Done

**Review by AatishBondge — COMMENTED** [review]
Looks goood for me, @bhushan-oe please review

File: client/src/actions/projectAction.js [hunk 18]
```diff
@@ -100,3 +114,88 @@ export function clearProjectMsg() {
...
+export function setProjectAllocationDataErr(data) {
+  return {
+    type: PROJECT_ALLOCATION_DATA_ERROR,
+    data
+  }
+}
+export function unassignProject(id, enddate) {
+  return {
+    type: DEALLOCATE_PROJECT,
+    payload: { id: id, enddate: enddate }
```
Comment (bhushan-oe) [inline client/src/actions/projectAction.js]: Use property shorthand

File: client/src/components/Project/updateProjectAllocation.js [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+    null
+  )
+  const projectId = projectToUpdate ? projectToUpdate[0]._id : null
+  const links = ['Deallocate', 'Delete']
+
+  useEffect(() => {
+    if (projectId) {
+      dispatch(getProjectAllocationData(projectId))
+    }
+  }, [projectId])
```
Comment (bhushan-oe) [inline client/src/components/Project/updateProjectAllocation.js]: dispatch as dependency

File: client/src/components/Project/updateProjectAllocation.js [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+  }, [projectAllocationDetailsErr])
+
+  useEffect(() => {
+    if (projectAllocationDetails) {
+      //to do sort data before display
+
+      const data = projectAllocationDetails.map(prop => {
+        let {
+          employee: { firstname },
+          employee: { lastname },
```
Comment (bhushan-oe) [inline client/src/components/Project/updateProjectAllocation.js]: overriding employee key

File: client/src/components/Project/updateProjectAllocation.js [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+          startdate,
+          enddate,
+          functional_manager: { firstname: mgr_firstname },
+          functional_manager: { lastname: mgr_lastname },
```
Comment (bhushan-oe) [inline client/src/components/Project/updateProjectAllocation.js]: overriding key

File: client/src/components/Project/updateProjectAllocation.js (hunk 20 above)
Comment (bhushan-oe) [reply client/src/components/Project/updateProjectAllocation.js]: `employee: {firstname, lastname} `

**Review by bhushan-oe — CHANGES_REQUESTED** [review]

File: client/src/actions/projectAction.js (hunk 18 above)
Comment (surekhag) [reply client/src/actions/projectAction.js]: Done.

File: client/src/components/Project/updateProjectAllocation.js (hunk 19 above)
Comment (surekhag) [reply client/src/components/Project/updateProjectAllocation.js]: Added.

File: client/src/components/Project/updateProjectAllocation.js (hunk 20 above)
Comment (surekhag) [reply client/src/components/Project/updateProjectAllocation.js]: Updated.

File: client/src/components/Project/updateProjectAllocation.js (hunk 21 above)
Comment (surekhag) [reply client/src/components/Project/updateProjectAllocation.js]: Updated.

**Review by bhushan-oe — APPROVED** [review]

**Pull request merged**